File: design_patterns/C++/Behavioral/mediator.cxx
#include <iostream>
#include <string>
#include <ctime>

class User;

class ChatRoomMediator {
  public:
	virtual void showMessage(User* user, std::string message) = 0;
};

class ChatRoom: public ChatRoomMediator {
  public:
	void showMessage(User* user, std::string message);
};

class User {
  protected:
	std::string name = "";
	ChatRoomMediator* chatMediator;

  public:
	User(std::string name, ChatRoomMediator* chatMediator) {
		this->name = name;
		this->chatMediator = chatMediator;
	}

	std::string getName() {
		return this->name;
	}

	void send(std::string message) {
		this->chatMediator->showMessage(this, message);
	}
};

void ChatRoom::showMessage(User* user, std::string message) {
	auto t = time(nullptr);
	auto charTime = ctime(&t);
	auto sender = user->getName();
	std::cout << charTime << "[" << sender << "]: " << message << std::endl;
}

int main() {
	auto mediator = new ChatRoom();
	auto john = new User("John Doe", mediator);
	auto jane = new User("Jane Doe", mediator);
	john->send("Hi there!");
	jane->send("Hey!");
	// ========================
	// Tue Feb 28 15:31:01 2017
	// [John Doe]: Hi there!
	// Tue Feb 28 15:31:01 2017
	// [Jane Doe]: Hey!
	// ========================
	return 0;
}

File: design_patterns/Python/Behavioral/mediator.py
from datetime import datetime


class ChatRoomMediator:
    pass


class ChatRoom(ChatRoomMediator):

    def show_message(self, user, message):
        time = datetime.now()
        sender = user.get_name()
        print('[{}] {}: {}'.format(time, sender, message))


class User:

    def __init__(self, name, chat_room_mediator):
        self.name = name
        self.chat_room_mediator = chat_room_mediator

    def get_name(self):
        return self.name

    def send(self, message):
        self.chat_room_mediator.show_message(self, message)


if __name__ == '__main__':

    mediator = ChatRoom()
    john = User('John Doe', mediator)
    jane = User('Jane Doe', mediator)
    john.send('Hi there!')
    jane.send('Hey!')
    # Output:
    # [2017-02-24 10:31:07.474657] John Doe: Hi there!
    # [2017-02-24 10:31:07.475182] Jane Doe: Hey!
